latch_cli/services/execute/main.py

```diff
@@ -0,0 +1,183 @@
+    data: str
+
+
+class CloseResponse(TypedDict):
+    stream: Literal["close"]
+
+
+Response: TypeAlias = Union[StdoutResponse, CloseResponse]
+
+
+async def get_stdio_streams():
+    loop = asyncio.get_event_loop()
+    reader = asyncio.StreamReader(loop=loop)
+
+    stdin2 = os.dup(sys.stdin.fileno())
+    stdout2 = os.dup(sys.stdout.fileno())
+
+    await loop.connect_read_pipe(
+        lambda: asyncio.StreamReaderProtocol(
+            reader,
+            loop=loop,
+        ),
+        os.fdopen(stdin2),
+    )
+
+    writer_transport, writer_protocol = await loop.connect_write_pipe(
+        lambda: asyncio.streams.FlowControlMixin(loop=loop),
+        os.fdopen(stdout2),
+    )
+    writer = asyncio.streams.StreamWriter(
+        writer_transport,
+        writer_protocol,
+        None,
+        loop,
+    )
+
+    return reader, writer
+
+
+async def shutdown_on_close(ws: websockets.WebSocketClientProtocol, task: asyncio.Task):
+    await ws.wait_closed()
+    task.cancel()
+
+
+async def send_stdin(
+    ws: websockets.WebSocketClientProtocol, reader: asyncio.StreamReader
+):
+    while True:
+        message = await reader.read(1)
+
+        await ws.send(json.dumps({"stream": "stdin", "data": message.decode("utf-8")}))
+
+
+async def write_stdout(
+    ws: websockets.WebSocketClientProtocol, writer: asyncio.StreamWriter
+):
+    async for output in ws:
+        if isinstance(output, bytes):
+            output = output.decode("utf-8")
+
+        try:
+            res = json.loads(output)
+            # todo(ayush): write stderr messages to stderr lol
+            if res["stream"] == "close":
+                break
+
+            writer.write(res["data"].encode())
+        except (json.JSONDecodeError, KeyError) as e:
+            # todo(ayush) error surfacing
+            break
+
+    await ws.close()
+
+
+async def handle_resize(resize_event_queue: asyncio.Queue[Tuple[int, int]]):
+    await resize_event_queue.put(os.get_terminal_size())
+
+
+async def propagate_resize_events(
+    ws: websockets.WebSocketClientProtocol,
+    resize_event_queue: asyncio.Queue[Tuple[int, int]],
+):
+    while True:
+        cols, rows = await resize_event_queue.get()
+
+        await ws.send(
+            json.dumps({
+                "stream": "resize",
+                "size": {
+                    "Width": cols,
+                    "Height": rows,
+                },
+            })
+        )
+
+
+async def connect(egn_info: EGNNode, container_info: Optional[ContainerNode]):
+    loop = asyncio.get_event_loop()
+
+    resize_event_queue: asyncio.Queue[tuple[int, int]] = asyncio.Queue()
+    await resize_event_queue.put(os.get_terminal_size())
+
+    loop.add_signal_handler(
+        signal.SIGWINCH,
+        lambda: asyncio.create_task(handle_resize(resize_event_queue)),
+    )
+
+    reader, writer = await get_stdio_streams()
+
+    async with websockets.connect(
+        urlparse(urljoin("http://localhost:5000", "/workflows/cli/shell"))
```

Comment:
`localhost:5000` is a bad choice
I'd do like `5554` (random number)

`5000` already has a daemon running on mac by default and is likely to conflict with other dev servers

Reply:
unless this is just for debugging? in which case this is just probably wrong

Reply:
just for debugging, will remove